### PR TITLE
SAK-44967 Allow students to see the grades of their classmates

### PR DESCRIPTION
**Pull request by brdebr**
Hi, I've uploaded some previews of this feature in the JIRA

### CODE REVIEW

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 1]
```diff
@@ -1018,7 +1018,7 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
 		List<GbStudentGradeInfo> items = sortGradeMatrix(matrix, settings);
 		stopwatch.timeWithContext("buildGradeMatrix", "sortGradeMatrix", stopwatch.getTime());
 
-		return items;
+        return items;
```
Comment (mpellicer) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:1021]: Remove this change

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 2]
```diff
@@ -1073,6 +1073,23 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
 
 		return items;
 	}
+    
+	public List<GbStudentGradeInfo> buildMatrixForGradeComparison(Assignment assignment){
+		// Only return the list if the feature is activated
+		boolean serverPropertyOn = serverConfigService.getConfig("gradebookng.allowStudentsToCompareGradesWithClassmates", false);
+		if(!serverPropertyOn){
```
Comment (mpellicer) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:1080]: spacing if (XXX) {

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 3]
```diff
@@ -1073,6 +1073,23 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+			return new ArrayList<>();
+		}
+
+		// Add advisor to retrieve the grades as student
+		SecurityAdvisor advisor = null;
+		try{
+			advisor = addSecurityAdvisor();
+			return buildGradeMatrix(Collections.singletonList(assignment));
+		}finally{
```
Comment (mpellicer) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:1089]: spacing } finally {

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 4]
```diff
@@ -1073,6 +1073,23 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+	public List<GbStudentGradeInfo> buildMatrixForGradeComparison(Assignment assignment){
+		// Only return the list if the feature is activated
+		boolean serverPropertyOn = serverConfigService.getConfig("gradebookng.allowStudentsToCompareGradesWithClassmates", false);
+		if(!serverPropertyOn){
+			return new ArrayList<>();
+		}
+
+		// Add advisor to retrieve the grades as student
+		SecurityAdvisor advisor = null;
+		try{
```
Comment (mpellicer) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:1086]: spacing try {

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 5]
```diff
@@ -1073,6 +1073,23 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
 
 		return items;
 	}
+    
+	public List<GbStudentGradeInfo> buildMatrixForGradeComparison(Assignment assignment){
+		// Only return the list if the feature is activated
+		boolean serverPropertyOn = serverConfigService.getConfig("gradebookng.allowStudentsToCompareGradesWithClassmates", false);
```
Comment (mpellicer) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:1079]: Is there a common property files in GradebookNG? If not no problem

File: gradebookng/tool/src/webapp/styles/gradebook-grades-comparison.css [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  display: flex;
+  align-content: center;
+  justify-content: center;
+}
+
+#loading-grade-summary img{
+  height: 20px;
+  width: 20px;
+  margin: auto;
+}
```
Comment (mpellicer) [inline gradebookng/tool/src/webapp/styles/gradebook-grades-comparison.css:49]: New line

**Review by mpellicer — CHANGES_REQUESTED** [review]
Well done overall @brdebr ! Some minor comments to address

**Comment by mpellicer** [comment]
Also please address the accessibility complains, if possible.

File: config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties [hunk 7]
```diff
@@ -1950,6 +1950,10 @@
 # DEFAULT: true
 # gradebookng.showDisplayCourseGradeToStudent=false
 
+# SAK-44967: control the ability for students to compare their grades with their classmates.
+# DEFAULT: false
+# gradebookng.allowStudentsToCompareGradesWithClassmates=false
```
[the remark this answers is not in the document]
Comment (brdebr) [reply config/configuration/bundles/src/bundle/org/sakaiproject/config/bundle/default.sakai.properties]: Done

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java [hunk 8]
```diff
@@ -89,5 +89,41 @@
...
 	private boolean courseGradeStatsDisplayed;
+    
+
+	/**
+	 * Are students allowed to compare grades with classmates?
+	 */
+	private boolean allowStudentsToCompareGrades;
+
+	/**
+	 * Are grades diplayed when comparing grades with classmates?
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java]: Typ , copied and pasted 😅

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java [hunk 9]
```diff
@@ -89,5 +89,41 @@
...
+	 */
+	private boolean comparingDisplayGrades;
+
+	/**
+	 * Are student names diplayed when comparing grades with classmates?
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java]: Typo , copied and pasted 😅

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java [hunk 10]
```diff
@@ -89,5 +89,41 @@
...
+	 */
+	private boolean comparingDisplayStudentNames;
+
+	/**
+	 * Are student surnames diplayed when comparing grades with classmates?
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java]: Typo , copied and pasted 😅

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java [hunk 11]
```diff
@@ -89,5 +89,41 @@
...
+	 */
+	private boolean comparingDisplayStudentSurnames;
+
+	/**
+	 * Are teacher comments diplayed when comparing grades with classmates?
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradebookInformation.java]: Typo , copied and pasted 😅

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 12]
```diff
@@ -465,6 +474,13 @@ public GradebookInformation getGradebookInformation(final String gradebookUid) {
 		gradebook.setCourseGradeStatsDisplayed(gradebookInformation.isCourseGradeStatsDisplayed());
 		gradebook.setAssignmentsDisplayed(gradebookInformation.isDisplayReleasedGradeItemsToStudents());
 		gradebook.setCourseGradeDisplayed(gradebookInformation.isCourseGradeDisplayed());
+        gradebook.setAllowStudentsToCompareGrades(gradebookInformation.isAllowStudentsToCompareGrades());
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: Done

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 13]
```diff
@@ -3301,6 +3317,15 @@ public void updateGradebookSettings(final String gradebookUid, final GradebookIn
 		// set stats display settings
 		gradebook.setAssignmentStatsDisplayed(gbInfo.isAssignmentStatsDisplayed());
 		gradebook.setCourseGradeStatsDisplayed(gbInfo.isCourseGradeStatsDisplayed());
+        
+        // set allow students to compare grades
```
[the remark this answers is not in the document]
Comment (brdebr) [reply edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 14]
```diff
@@ -970,7 +970,7 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
 	 */
 	public List<GbStudentGradeInfo> buildGradeMatrix(final List<Assignment> assignments,
 			final List<String> studentUuids, final GradebookUiSettings uiSettings) throws GbException {
-
```
[the remark this answers is not in the document]
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 1 above)
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 2 above)
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 3 above)
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 4 above)
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Done

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 5 above)
Comment (brdebr) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: I've moved it to the top of the file, but there's no place where the properties are centralized

File: gradebookng/tool/src/webapp/styles/gradebook-grades-comparison.css (hunk 6 above)
Comment (brdebr) [reply gradebookng/tool/src/webapp/styles/gradebook-grades-comparison.css]: Done

**Comment by brdebr** [comment]
About the acceslint issues,
I think they are false positives, 
those inputs are labelled implicitly by the wrapping <label>

https://www.w3.org/WAI/tutorials/forms/labels/#associating-labels-implicitly

**Comment by mpellicer** [comment]
Thanks @brdebr, good job so far.

This feature was presented in a T&L call by @danielmerino and it was majorly accepted, even they made suggestions that were implemented. The implementation looks correct to me, it's disabled by default and invisible for the end user unless it's enabled.

@juanjmerono and his team are interested in testing this feature, also @danielmerino and @vacercon are waiting for this change to be available for testing in Nightly, we're going to enable it in the experimental server.

If anyone has later concerns no problem on raising related Jiras or use the same Jira number.

**Review by mpellicer — APPROVED** [review]

**Comment by mpellicer** [comment]
This is paired with https://github.com/sakaiproject/sakai-reference/pull/121

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+                    <tbody>
+                        <tr wicket:id="gradeItemRows">
+                            <td wicket:id="studentNameLabel">
+                                Student name value
+                            </td>
+                            <td wicket:id="teacherCommentLabel">
+                                Teacher comment value
+                            </td>
+                            <td wicket:id="gradeLabel" class="text-center">
+                                Grade value
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:37]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+                            Grade
+                        </th>
+                    </thead>
+                    <tbody>
+                        <tr wicket:id="gradeItemRows">
+                            <td wicket:id="studentNameLabel">
+                                Student name value
+                            </td>
+                            <td wicket:id="teacherCommentLabel">
+                                Teacher comment value
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:34]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+                            Teacher comments
+                        </th>
+                        <th wicket:id="gradeHeaderLabel">
+                            Grade
+                        </th>
+                    </thead>
+                    <tbody>
+                        <tr wicket:id="gradeItemRows">
+                            <td wicket:id="studentNameLabel">
+                                Student name value
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:31]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+                <table class="table table-bordered table-condensed table-hover">
+                    <thead>
+                        <th wicket:id="studentNameHeaderLabel">
+                            Student Name
+                        </th>
+                        <th wicket:id="teacherCommentHeaderLabel">
+                            Teacher comments
+                        </th>
+                        <th wicket:id="gradeHeaderLabel">
+                            Grade
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:25]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+                </span>
+            </div>
+            <div class="studentGradesComparison-TableContainer">
+                <table class="table table-bordered table-condensed table-hover">
+                    <thead>
+                        <th wicket:id="studentNameHeaderLabel">
+                            Student Name
+                        </th>
+                        <th wicket:id="teacherCommentHeaderLabel">
+                            Teacher comments
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:22]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+                </span>
+                <span wicket:id="gradeItemLabel">
+                    Item name
+                </span>
+            </div>
+            <div class="studentGradesComparison-TableContainer">
+                <table class="table table-bordered table-condensed table-hover">
+                    <thead>
+                        <th wicket:id="studentNameHeaderLabel">
+                            Student Name
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:19]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+<body>
+    <wicket:panel>
+        <div id="studentGradesComparison-ModalContent">
+            <div class="studentGradesComparison-Header">
+                <span>
+                    <wicket:message key="comparegrades.modal.subtitle.label" />
+                </span>
+                <span wicket:id="gradeItemLabel">
+                    Item name
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:12]: Hardcoded

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 22]
```diff
@@ -188,6 +188,9 @@
 	public static final String ICON_SAKAI = "icon-sakai--";
 	public static final String ALL = "all";
 
+	private static final String SAK_PROP_ALLOW_STUDENTS_TO_COMPARE_GRADES = "gradebookng.allowStudentsToCompareGradesWithClassmates";
+	private static final Boolean SAK_PROP_ALLOW_STUDENTS_TO_COMPARE_GRADES_DEFAULT = false;
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:192]: Boolean.FALSE;

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 15 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:37]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 16 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:34]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 18 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:25]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 19 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:22]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 20 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:19]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 21 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:12]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 22 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:192]: I don't see why `Boolean.FALSE` is necessary here.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html (hunk 17 above)
Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:31]: This is a wicket tag, and the hard coded value here will be replaced by the wicket message specified in Java. This is a non-issue.

Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/StudentCompareGradesPanel.html:31]: Thanks @bjones86

**Pull request merged**